File: Stock Buy and Sell – Max one Transaction Allowed.java
class Solution {
    public int maximumProfit(int prices[]) {
        // Code here
        if (prices == null || prices.length <= 1) {
            return 0;
        }
        
        int minPrice = Integer.MAX_VALUE;
        int maxProfit = 0;
        
        for (int price : prices) {
            if (price < minPrice) {
                minPrice = price;
            }
            int profit = price - minPrice;
            if (profit > maxProfit) {
                maxProfit = profit;
            }
        }
        
        return maxProfit;
    }
}

File: Stock Buy and Sell – Max one Transaction Allowed.cpp
class Solution {
  public:
    int maxProfit(vector<int> &p) {
        int n = p.size(),maxi=0;
        vector<int>v(n,0);
        for(int i=n-1;i>=0;--i){
            if(i==n-1) v[i] = p[i];
            else{
                v[i]=max(p[i],v[i+1]);
            }
            maxi=max(maxi,v[i]-p[i]);
        }
        return maxi;
        
    }
};
